File: leetcode/0491-increasing-subsequences/solution.cpp
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        backtrack(nums, 0);
        return result;
    }

private:
    vector<vector<int>> result;
    vector<int> path;

    void backtrack(vector<int>& nums, int start) {
        if (path.size() >= 2) {
            result.push_back(path);
        }

        unordered_set<int> seenThisLevel;
        for (int i = start; i < (int)nums.size(); i++) {
            if (seenThisLevel.count(nums[i])) {
                continue;
            }
            if (!path.empty() && nums[i] < path.back()) {
                continue;
            }
            seenThisLevel.insert(nums[i]);
            path.push_back(nums[i]);
            backtrack(nums, i + 1);
            path.pop_back();
        }
    }
};

File: leetcode/0491-increasing-subsequences/solution.py
from typing import List


class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        result = []
        path = []
        n = len(nums)

        def backtrack(start):
            if len(path) >= 2:
                result.append(path[:])

            seen_this_level = set()
            for i in range(start, n):
                if nums[i] in seen_this_level:
                    continue
                if path and nums[i] < path[-1]:
                    continue
                seen_this_level.add(nums[i])
                path.append(nums[i])
                backtrack(i + 1)
                path.pop()

        backtrack(0)
        return result
